week8/UniversityStore.swift

```diff
@@ -0,0 +1,77 @@
+//
+//  UniversityStore.swift
+//  Mod8JSON
+//
+//  Created by Nick Ryan on 6/9/24.
+//
+
+import Foundation
+import SwiftUI
+
+class UniversityStore: ObservableObject {
+  @Published var universities: [UniversityEntry] = []
+  @Published var downloadProgress: Float = 0
+  
+  private let decoder = JSONDecoder()
+  private let session: URLSession
+  private let sessionConfirguration: URLSessionConfiguration
+  
+  init() {
+    self.sessionConfirguration = URLSessionConfiguration.default
+    self.session = URLSession(configuration: sessionConfirguration)
+  }
+  
+  func fetchDomain() async throws {
+    
+    guard let url = URL(string:  "http://universities.hipolabs.com/search?country=United+States")
+    else {
+      print("Invalid URL")
+      return
+    }
+    
+    let request = URLRequest(url: url)
+    let (asyncBytes, response) = try await session.bytes(for: request)
+    
+    guard let httpResponse = response as? HTTPURLResponse,
+          httpResponse.statusCode == 200
+    else {
+      print("Invalid response code.")
+      
+      return
+    }
+    
+    let contentLength = Float(response.expectedContentLength)
+    var data = Data(capacity: Int(contentLength))
+    
+    for try await byte in asyncBytes {
+      data.append(byte)
+      
+      let currentProgress = Float(data.count) / contentLength
+      
+      DispatchQueue.main.async {
+        self.downloadProgress = currentProgress
+      }
```

Comment:
Instead of using this here, you should make this async method run on the main actor like so:
```
@MainActor func fetchDomain() async throws {
```
That way, you don't need to use DispatchQueue. The code inside this method will run on the main thread (asynchronous) but all the rest that is in an `await` call will run on some other thread.